### PR TITLE
Feauture/eligible moves

### PR DESCRIPTION
**Pull request by kpesaran**
Added logic to eligibleMoves method to find allowed positions a robot is allowed to move into given a direction. The method returns a position based on board size, if a cell is obstructed, if another robot is encountered, and if a wall obstructs an exit from a cell.

### CODE REVIEW

File: frontend/src/util/boardBuilder.test.ts [hunk 1]
```diff
@@ -35,27 +35,27 @@ test(
...
   () => {
     let board = new BoardBuilder()
-      .withWall(Wall.North, { row: 0, column: 0 })
-      .withWall(Wall.West, { row: 0, column: 0 })
-      .withWall(Wall.East, { row: 15, column: 15 })
-      .withWall(Wall.South, { row: 15, column: 15 })
+      .withWall(Direction.North, { row: 0, column: 0 })
+      .withWall(Direction.West, { row: 0, column: 0 })
+      .withWall(Direction.East, { row: 15, column: 15 })
+      .withWall(Direction.South, { row: 15, column: 15 })
```
Comment (mitchpowell1) [inline frontend/src/util/boardBuilder.test.ts:55]: nit:

Ideally the Wall -> Direction changes would be moved to their own pull request to make the functional changes of this pull request less ambiguous.

I know we talked about it last week and I said not to worry about it, just something to think about :)

File: frontend/src/main.js [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
-import { placeWalls, onMouseDown,onMouseMove} from './utilites.js'
+import { placeWalls, onMouseDown, onMouseMove } from './utilites.js'
+import { BoardBuilder } from './util/boardBuilder.ts'
 import * as THREE from 'three'
 
-
-const boardPositions = new Positions
-
+const boardBuilder = new BoardBuilder()
+console.log(boardBuilder)
```
Comment (mitchpowell1) [inline frontend/src/main.js:12]: nit

log statements intended for debugging should be removed

File: frontend/src/board/board.ts [hunk 3]
```diff
@@ -45,5 +47,105 @@ export class Board {
       { row: 0, column: 2},
       { row: 0, column: 3},
     ];
+
+  }
+
+  findMoves() {
+    return {
+      north: this.checkDirections(this.robotPositions[0].row, this.robotPositions[0].column, 'North'),
```
Comment (mitchpowell1) [inline frontend/src/board/board.ts]: Indentation looks a little funny here, I think we need to move south, east, west to this indentation level as well

File: frontend/src/board/board.ts [hunk 4]
```diff
@@ -45,5 +47,105 @@ export class Board {
...
+    south: this.checkDirections(this.robotPositions[0].row, this.robotPositions[0].column, 'South'),
+    east: this.checkDirections(this.robotPositions[0].row, this.robotPositions[0].column, 'East'),
+    west: this.checkDirections(this.robotPositions[0].row, this.robotPositions[0].column, 'West')
+    }
+  }
+
+  checkDirections(startRow:number, startCol: number, direction: 'North' | 'South' | 'East' | 'West' ) : null | Position {
```
Comment (mitchpowell1) [inline frontend/src/board/board.ts]: Can we use our `Direction` enum here? Also maybe a `Position` as well

File: frontend/src/board/board.ts [hunk 5]
```diff
@@ -45,5 +47,105 @@ export class Board {
...
+    while (row >= 0 && row < BOARD_SIZE && col >= 0 && col < BOARD_SIZE) {
+
+      // check if target, add position
+
+      let robotEncountered = false 
+      // robots 
+      this.robotPositions.forEach(robot => {
+        // designated robot does not check its own position
+        if (robot.row === startRow && robot.column === startCol) {
+          return
```
Comment (mitchpowell1) [inline frontend/src/board/board.ts:83]: This check is probably unnecessary - there should only be on robot per square (and if there isn't we probably have something we need to fix somewhere else)

File: frontend/src/board/board.ts [hunk 6]
```diff
@@ -45,5 +47,105 @@ export class Board {
...
+
+      
+      // check if obstructed
+      if (this.cells[row]?.[col]?.isObstructed) {
+        break
+      }
+
+      let wallPreventsEntrance = false 
+      // check if wall exists
+      if (this.cells[row]?.[col]?.walls?.length! > 0) {
```
Comment (mitchpowell1) [inline frontend/src/board/board.ts:101]: I think we can remove this condition

File: frontend/src/board/board.ts [hunk 7]
```diff
@@ -45,5 +47,105 @@ export class Board {
...
+            legalMove = {row: row, column:col}
+          } 
+          
+    
+        })
+      }
+      if (wallPreventsEntrance) {
+        
+        break
+      }
```
Comment (mitchpowell1) [inline frontend/src/board/board.ts]: We can probably simplify this a bit:

```ts
if this.cells[row]?.[col]?.walls.some(wall => {
// check walls
}) {
  break;
}
```

File: frontend/src/board/board.ts [hunk 8]
```diff
@@ -45,5 +47,105 @@ export class Board {
...
+
+      if (direction === 'North') {
+        row--;
+      } else if (direction === 'South') {
+        row++;
+      } else if (direction === 'East') {
+        col++;
+      } else if (direction === 'West') {
+        col--;
+      }
```
Comment (mitchpowell1) [inline frontend/src/board/board.ts]: As a broad meta level comment I think it might simplify things a little bit if we increment our position before entering the while loop so we can always assume that our current `(row, col)` values are a valid legal move, and break if we discover that the next coordinates are not. Let me know if that doesn't make sense

File: frontend/src/board/board.test.ts [hunk 9]
```diff
@@ -46,3 +47,93 @@ test(
...
+test('Checking eligible move in the West direction', () => {
+  let board = new Board();
+  expect(board.checkDirections(6, 15, 'West')).toStrictEqual(
+  {row: 6, column: 0})
+}
+)
+
+test("Should return target cell in path", () => {
+  let board = new Board();
+  if (board.cells[15] && board.cells[15][14]) {
```
Comment (mitchpowell1) [inline frontend/src/board/board.test.ts]: This condition should always be true - and it's alright if we crash in a unit test, that's a valuable data point

Can we just do `board.cells[15]!.[14]!.isTarget = true;`?

File: frontend/src/board/board.ts [hunk 10]
```diff
@@ -45,5 +47,105 @@ export class Board {
...
+
+      if (robotEncountered) {
+        break; 
+      }
+
+
+      if (this.cells[row]?.[col]?.isTarget) {
+        legalMove = {row: row, column: col}
+        break
+      }
```
Comment (mitchpowell1) [inline frontend/src/board/board.ts]: I think this is a logic error - the robot will slide right over the target if there isn't stopping it from doing so

File: frontend/src/board/board.test.ts [hunk 11]
```diff
@@ -46,3 +47,93 @@ test(
...
+    board.cells[15][14].isTarget = true;
+  }
+  expect(board.checkDirections(9,14,"South")).toStrictEqual({row:15,column:14})
+  expect(board.checkDirections(15,9,"East")).toStrictEqual({row:15,column:14})
```
Comment (mitchpowell1) [inline frontend/src/board/board.test.ts]: See comments in `board.ts` but I don't think this is actually the behavior we want - the robot shouldn't stop sliding unless it runs into something

File: frontend/src/board/board.test.ts [hunk 12]
```diff
@@ -46,3 +47,93 @@ test(
...
+  expect(board.checkDirections(1, 1, 'South')).toStrictEqual(null)
+  expect(board.checkDirections(1, 1, 'East')).toStrictEqual({row:1, column: 15})
+
+
+  expect(board.checkDirections(5, 1, 'North')).toStrictEqual({row:2, column: 1})
+  expect(board.checkDirections(1, 2, 'West')).toStrictEqual({ row: 1, column: 1 })
+  expect(board.checkDirections(1, 8, 'West')).toStrictEqual({ row: 1, column: 1 })
+  expect(board.checkDirections(0, 1, 'South')).toStrictEqual({ row: 1, column: 1 })
+ 
+
```
Comment (mitchpowell1) [inline frontend/src/board/board.test.ts]: nit:

Formatting, we can delete these extra blank lines

**Review by mitchpowell1 — APPROVED** [review]
Left some comments for your consideration - I think that we should fix the bit where the robot stops sliding if it hits the target since that isn't the behavior we want, but otherwise looks good!

**Comment by kpesaran** [comment]
Thank you, I made some changes based on your comments and pushed them on this branch.
Changes include:
- Removing any code, and corresponding testing, that stopped the robot from sliding past the target cell
- Refactored The checkDirections method, and corresponding testing, to use the Direction enum and Position types.
